src/Controller/DBHandler.js

```diff
@@ -0,0 +1,450 @@
+      }
+    } catch (error) {
+      if (isGoHome) {
+        this.setGoHome(true);
+      }
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //get the user Info
+  getUserInfo = async (isGoHome) => {
+    const userInfoRef = ref(database, `userInfo/${this.uid}`);
+    try {
+      const newUserInfo = await get(userInfoRef);
+      return newUserInfo.val();
+    } catch (error) {
+      if (isGoHome) {
+        this.setGoHome(true);
+      }
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //get the user quiz list
+  getUserQuizList = async () => {
+    const quizListRef = ref(database, `userInfo/${this.uid}/quizReport`);
+    try {
+      const res = await get(quizListRef);
+      const quizList = res.val();
+      if (!quizList) {
+        throw new Error("You have no Quiz taken.");
+      }
+      return quizList;
+    } catch (error) {
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //get the user quiz report
+  getUserQuizReport = async (quizNo) => {
+    const quizRef = ref(
+      database,
+      `userInfo/${this.uid}/quizReport/quiz${quizNo}`
+    );
+    try {
+      const res = await get(quizRef);
+      const quizReport = res.val();
+      if (!quizReport) {
+        throw new Error("You don't have this quiz.");
+      }
+      return quizReport;
+    } catch (error) {
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //get single deck Info
+  getDeckInfo = async (deckID, isGoHome) => {
+    try {
+      const decksRef = ref(database, `decks/deck${deckID}`);
+      const res = await get(decksRef);
+      return res.val();
+    } catch (error) {
+      if (isGoHome) {
+        this.setGoHome(true);
+      }
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //get single card Info
+  getCardInfo = async (cardID, isGoHome) => {
+    try {
+      const cardRef = ref(database, `cards/card${cardID}`);
+      const res = await get(cardRef);
+      return res.val();
+    } catch (error) {
+      if (isGoHome) {
+        this.setGoHome(true);
+      }
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //Get single deck and cards in that deck
+  //return { deckInfo, cardsInfo }
+  getDeckAndCards = async (deckID, isGoHome) => {
+    try {
+      const deckInfo = await this.getDeckInfo(deckID, isGoHome);
+      if (!deckInfo) throw new Error("Deck not found");
+
+      const cardNumber = Object.values(deckInfo.deckCards);
+      const cardPromises = cardNumber.map(
+        async (cardID) => await this.getCardInfo(cardID, isGoHome)
+      );
+      const cardsInfo = await Promise.all(cardPromises);
+      return { deckInfo, cardsInfo };
+    } catch (error) {
+      if (isGoHome) {
+        this.setGoHome(true);
+      }
+      this.setErrorMessage(error.message);
+    }
+  };
+
+  //search Deck with user doesn't have.
+  //return {resultInfo,adviceInfo] (array of deckInfo with new key cardInfos contains cards detail in that deck)
+  searchData = async (keywordRaw) => {
```

Comment:
This function could use a lot of abstraction into functions that perform sub-tasks here